### DISCUSSION TITLE
var, let, const에 대해서 설명해주세요

### DISCUSSION THREAD

**Discussion by kamwoo** [woowacourse-study/2022-woowahan-bansanghwe#98] [JavaScript]
## 변수란

> 데이터에 식별자를 명시한것
> 
- 데이터를 일정 기간 동안 저장하여 필요한 때에 다시 사용하기 위해 사용된다.
- 사람이 이해하기 쉽게 어떠한 값의 메모리 주소에 접근하기 위해서 사용한다.
- 식별자에는 변수명, 함수명, 프로퍼티명, 클래스명 등이 있다.

## var, let, const

변수는 `var`, `let`, `const`를 사용해서 선언하고, 값을 할당한다.

### var

ES6 이전까지 `var`를 사용해서 변수를 선언했다. 하지만 `var`의 특징으로 인해서 단점이 있었고 자바스크립트에서 되도록 사용하지 않는 방향으로 가이드하고 있다.

특징

1. 변수 중복 선언 가능
    
    : 의도하지 않은 변수값의 변경이 일어날 가능성이 크다.
    
2. 함수 레벨 스코프
    
    : 함수의 코드 블럭만을 스코프로 인정한다. 함수에서만 지역 스코프를 가진다.
    
3. 변수 호이스팅
    
    : 변수를 선언하기 이전에 참조할 수 있다.
    

→ 전반적으로 코드가 어떻게 작동될지 예측하기 어려운 경우가 자주 발생한다.

### const, let

- let, const는 함수와 if, for, while, try/catch 등 블록레벨 스코프를 가진다.
- let으로 선언하면 초기화할 때 undefined가 들어가고 넘어가지만
    
    const로 선언하면 undefined가 들어가고 초기화 시킬 값이 없으므로 초기값이 없다는 에러가 뜬다.
    

## 예시

```jsx
var a = 10; 
let b = 20; 
const c = 30; 

// 함수 선언문 
function func_1(arg){ return arg; } 

// 함수 표현식(var) 
var func_2 = function(arg){ return arg; } 

// 함수 표현식(const/let) 
const func_3= function(arg){ return arg; }

```

스캔

1. var
    
    컨텍스트에 var a 식별자를 저장하면서 바로 undefined를 할당한다.
    
    그래서 var a 선언문 전에 콘솔을 찍으면, 값은 undefined로 나오게 된다.
    
2. let, const
    
    컨텍스트에 let b, const c 식별자를 저장만 한다.
    
    그래서 선언문 전에 콘솔을 찍으면, reference error가 발생한다.
    
3. function
    - 함수 선언문
        1. 컨텍스트에 func_1 식별자를 저장한다.
        2. 힙에 메모리 공간을 확보한다.
        3. 콜스택에 힙 메모리 공간 주소를 할당한다.
        4. func_1 식별자에 콜스택 메모리 주소를 저장한다.
        5. 힙 메모리 공간에 함수 내용을 저장한다.
        
        그래서 함수 선언문 앞에서도 호출을 할 수 있다.
        
    - 함수 표현식
        
        변수로 선언한 것과 동일한 과정을 가진다.
        

실행

1. var
    
    var a = 10에 콜스택에 메모리 공간을 확보하여 10을 저장하고, 주소를 새로 저장한다.
    
2. let, const
    
    let b = 20, const c = 30을 만나면 undefined를 할당하고 메모리 공간을 확보해서 20, 30을 할당하여 주소를 저장한다.
    
3. function
    - 함수 선언문
        
        이미 스캔과정에서 메모리 할당까지 되었다.
        
    - 함수 표현식
        
        변수와 동일하다.

**Comment by JUDONGHYEOK** [comment]
자바스크립트에서 변수란?

**Comment by JUDONGHYEOK** [comment]
js에서 var, let, const 각각에 대해 설명 차이 자세히 설명

**Comment by JUDONGHYEOK** [comment]
왜 let, const를 사용하고 var를 이용하지 않을까?

**Comment by JUDONGHYEOK** [comment]
왜 자바스크립트는 함수레벨 스코프를 따랐을까요?

**Comment by JUDONGHYEOK** [comment]
var를 사용하지 않고 변수를 선언했을때?

**Comment by JUDONGHYEOK** [comment]
var let const가 실행컨텍스트에서 어떻게 동작하는지

**Comment by JUDONGHYEOK** [comment]
함수 선언문과 표현식의 차이는?

**Comment by JUDONGHYEOK** [comment]
호이스팅의 장단점?

**Comment by JUDONGHYEOK** [comment]
변수 호이스팅과 함수 호이스팅의 차이는?

**Comment by JUDONGHYEOK** [comment]
함수 표현식과 함수 선언의 호이스팅 차이는?

**Comment by JUDONGHYEOK** [comment]
JS의 콜스택과 메모리 힙에 대해 설명

**Comment by JUDONGHYEOK** [comment]
콜스택과 메모리힙을 나누는 이유는?

**Comment by JUDONGHYEOK** [comment]
크기가 고정된것과 고정되지 않은것은 js에서 뭐라고 부르나?

**Comment by JUDONGHYEOK** [comment]
참조타입의 메모리 할당 과정

**Comment by JUDONGHYEOK** [comment]
자바스크립트 배열에 대해서 설명해주세요

**Comment by JUDONGHYEOK** [comment]
C나 JAVA는 정적으로 분석해서 하나의 타입만 들어가는데, 동적타입인 JS에서는 어떻게 이를 계산해서 넣을까?

**Comment by JUDONGHYEOK** [comment]
자바스크립트에서 array를 변수로 선언하면 array를 메소드로 쓸떄를 사용할 수 있는데 그것이 가능한 이유?

**Comment by JUDONGHYEOK** [comment]
프로토타입으로부터 메소드를 사용할 수 있게 되는 과정
